### PR TITLE
CA-352988: Allow local live migration even when host is disabled

### PR DESCRIPTION
**Pull request by minglumlu**
Signed-off-by: Ming Lu <ming.lu@citrix.com>

### CODE REVIEW

File: ocaml/xapi/xapi_vm_helpers.ml [hunk 1]
```diff
@@ -661,8 +661,15 @@ let assert_can_boot_here ~__context ~self ~host ~snapshot ~do_cpuid_check
   validate_basic_parameters ~__context ~self ~snapshot ;
   assert_host_is_live ~__context ~host ;
   assert_matches_control_domain_affinity ~__context ~self ~host ;
+  let is_local_live_migration =
+    if (Db.VM.get_power_state ~__context ~self) = `Running
+       && (Db.VM.get_resident_on ~__context ~self) = host then
+      true
+    else
+      false
```
Comment (psafont) [inline ocaml/xapi/xapi_vm_helpers.ml:669]:
```suggestion
    (Db.VM.get_power_state ~__context ~self) = `Running
       && (Db.VM.get_resident_on ~__context ~self) = host
```

**Comment by lippirk** [comment]
If a host is disabled, is that not most likely because it is broken and the master cannot contact it?

Therefore doesn't removing this check just move the problem somewhere else?

**Comment by robhoes** [comment]
Allowing a VM power-state operation on a disabled host contradicts the meaning of enabled/disabled. You haven't explained on the PR why this is useful. In your scenario, can we not just re-enable the host before doing this localhost migration?

**Comment by minglumlu** [comment]
> Allowing a VM power-state operation on a disabled host contradicts the meaning of enabled/disabled. You haven't explained on the PR why this is useful. In your scenario, can we not just re-enable the host before doing this localhost migration?

Basically this host is still in updating. Keeping it disabled would protect this host from new VM start from outside.

**Comment by minglumlu** [comment]
> If a host is disabled, is that not most likely because it is broken and the master cannot contact it?
> 
> Therefore doesn't removing this check just move the problem somewhere else?
This change only allows a running VM on a disabled host to migrate on the same host. Indeed this allows a little bit more than before. But I don't think it breaks the meaning of disabled too much. The disabled host is still in protect basically.

**Comment by psafont** [comment]
I would be really helpful this information would be available in the commit message as well :)

**Comment by robhoes** [comment]
Yes, @minglumlu please update the commit message.

**Review by lindig — APPROVED** [review]

**Pull request merged**